### ISSUE TITLE
ImpactFlow

### ISSUE THREAD

**Issue by TSAMBALI** [TSAMBALI/project#113]
# ImpactFlow: Autonomous Social Good Architect

## Inspiration
ImpactFlow was inspired by the "Protocol" of social change—the idea that community problem-solving should be as rigorous and evidence-based as digital forensics. Witnessing the gap between young innovators (particularly Girls in STEM) and the technical complexity of modern software architecture, we built an agent that doesn't just "chat," but **architects**. 

We wanted to teach the agent how a senior systems analyst thinks: sequencing the approach, recognizing gaps in community logic, and autonomously adjusting parameters until a feasible, high-impact blueprint is forged.

## What we learned
We learned that **Autonomous Execution** in a social context requires strict guardrails. Using Gemini 1.5 Flash, we discovered that the agent's ability to self-correct during blueprint generation (recognizing when a tech stack is too complex for a high-schooler vs. a pro) is the "Screencast of live terminal execution" equivalent for social good.

## How we built it
The project is built on a full-stack architecture:
- **Frontend:** React 18 with Vite, styled with a "Vibrant Palette" using Tailwind CSS and Motion.
- **Backend:** Node.js/Express server acting as the secure agentic loop.
- **Intelligence:** Google Gemini 2.0 Flash Pro for the primary analytical loop.
- **Visualization:** Mermaid.js for autonomous architecture diagramming.
- **Tutorials:** React Joyride for a guided "SIFT-style" onboarding.

## Challenges we faced
The primary challenge was **Evidence Integrity**—ensuring the AI doesn't hallucinate "fake" solutions that aren't technically feasible. We implemented architectural guardrails where the agent must evaluate its own output against a "Complexity Score" (1-5) before presenting it to the user.

## Problem Statement
Traditional hackathons often suffer from "Ideation Bloat"—plenty of ideas, but zero technical depth. Middle and high school students often find the "last mile" of implementation (database schemas, API logic, middleware) a brick wall.

## Solution
ImpactFlow is an autonomous agent that bridges this gap. It takes a raw, messy community problem and applies a structured "SIFT" (Strategic Impact Framework Toolbox) process to extract a technically sound implementation blueprint.

## Tech Stack
- **Framework:** React + TypeScript
- **Styling:** Tailwind CSS (Vibrant Theme)
- **Agent Loop:** Express + @google/genai
- **Diagramming:** Mermaid.js
- **Animations:** Motion

## Future Scalability
- **Multi-Agent Track:** Decomposing blueprints into specialized sub-agents (one for UI/UX, one for Security, one for Database).
- **Persistent Learning:** Storing user feedback on blueprints to fine-tune the agent's "Senior Analyst" persona.
- **MCP Integration:** Allowing the agent to "write" initial code scaffolding directly via specialized servers.

---
*Created for the Girls in STEM Hackathon 2026. Empowering the next generation of technologists.*

<img width="1601" height="801" alt="Image" src="https://github.com/user-attachments/assets/e9928e46-1418-4396-ab04-a58e1b5ebda5" />

# PROJECT-14: Agent Capabilities & Target Users

## Problem Statement
What challenge does your agent solve?
It solves the **Accessibility Challenge** in software architecture. Most architecture tools are built for pros; ImpactFlow is built for the *next generation* of pros.

## Solution Overview
The agent addresses this by applying a **Forensic Mindset** to social issues. "Why does this problem exist?" -> "What evidence supports the fix?" -> "How do we build it without spoliation of the original intent?"

## Key Features
- **Autonomous Complexity Scoring:** Evaluates the difficulty level of a requested fix.
- **Tech Stack Harmonization:** Matches libraries to the user's skill level.
- **Real-Time Collaboration Lounge:** A dedicated space for team messaging and milestone tracking.

## Technologies Used
Implemented with React, Express, Gemini 2.0, and Motion for a polished "Vibrant" UI.

## Target Users
Junior technologists building for social good.

# PROJECT-2: Code Repository

**Link:** [https://github.com/aistudio-build/impact-flow](https://github.com/aistudio-build/impact-flow) (Public)
**License:** MIT License

The repository includes:
- `/src`: Modular React components.
- `/server.ts`: The secure agentic loop and API definitions.
- `package.json`: Clear dependency management for reproducibility.
- `metadata.json`: System-level capabilities declaration.

# PROJECT-3: Demo Video & Screencast

**Video URL:** [Link to 5 min screencast]

**Narrative Highlights:**
1. **The Autonomous Loop:** Showing the agent taking a vague "Poor lighting in parks" problem and self-correcting to include IoT motion sensors when it identifies a "Safety" category.
2. **Architecture Visualization:** The live generation of Mermaid.js diagrams based on the autonomous analysis.
3. **Evidence Integrity:** Showing the API logs in the terminal, demonstrating how the agent parses tokens and maintains context during the multi-turn generation.

# PROJECT-4: Architecture Diagram & Integrity

## Component Map
ImpactFlow follows a **Direct Agent Extension** pattern:
- **User Interface:** React components capturing the "Community Problem."
- **Agent Loop (Backend):** Express middleware that enforces architectural guardrails before the LLM returns content.
- **Data Pipeline:** Input -> Domain Extraction -> Complexity Ranking -> Blueprint Generation -> Diagram Rendering.

## Security vs. Prompt Guardrails
- **Prompt Guardrails:** The agent is instructed to use "Student Tips" and avoid destructive tech suggestions.
- **Architectural Guardrails:** The system enforces a `responseMimeType: "application/json"` to prevent the model from escaping its structured output, ensuring the data returned to the UI is always parsable and logic-consistent.

# PROJECT-5: Dataset Documentation

## Data Sources
1. **Gemini 2.0 Flash Training Set:** Provides the base logic for architecture and social impact metrics.
2. **Community Problem Corpus:** A curated list of 500+ common school and neighborhood issues used for testing extraction accuracy.
3. **Mermaid.js Pattern Library:** Structured templates for architectural flows.

## Reproducibility
The agent's findings can be reproduced by entering the exact strings into the `Analyze` input. Because the temperature is set to 0.2, the architectural decisions (Tech Stack, Phases) are highly deterministic and consistent across runs.

# PROJECT-7: Try-It-Out Instructions (Judges' Guide)

## Live Deployment
**URL:** [https://ais-dev-5esqxgrlvpxbeeq764o4si-143834394167.europe-west3.run.app](https://ais-dev-5esqxgrlvpxbeeq764o4si-143834394167.europe-west3.run.app)

## Step-by-Step Test Walkthrough
1. **Step 1:** Click "Watch Demo" or follow the **Interactive Tutorial** (it will auto-start for new users).
2. **Step 2:** Enter a problem: *"Our school cafeteria wastes 40% of its food daily."*
3. **Step 3:** Click **Analyze Problem**.
4. **Step 4:** Review the **Core Insights** and select the *"Cafeteria-to-Compost IoT Tracker"* project idea.
5. **Step 5:** Watch the **Mermaid Architecture** generate in real-time.
6. **Step 6:** Navigate to the **Pitch Coach** to see your narrative strategy.
7. **Step 7:** Visit the **Team Lounge** to simulate collaboration.

## Local Setup
1. `npm install`
2. Set `GEMINI_API_KEY` in your `.env`.
3. `npm run dev`

# PROJECT-8: Agent Execution Logs

## Trace Log (Sample)
```json
[
  {
    "timestamp": "2026-05-14T14:04:04Z",
    "event": "INPUT_RECEIVED",
    "token_count": 42,
    "payload": "Food waste in schools"
  },
  {
    "timestamp": "2026-05-14T14:04:06Z",
    "event": "DOMAIN_EXTRACTION",
    "result": ["Sustainability", "Agriculture"],
    "autonomous_correction": "Added Health category after finding latent nutritional impact"
  },
  {
    "timestamp": "2026-05-14T14:04:08Z",
    "event": "BLUEPRINT_GENERATED",
    "token_usage": 1420,
    "architecture_pattern": "Client-Server with Mermaid rendering"
  }
]
```

These traces allow judges to see how the agent "thinks" over time, ensuring every finding is backed by a specific tool execution.

# PROJECT-9: Agent Description

## Problem Statement
The challenge ImpactFlow solves is the **"Implementation Gap"** in social hackathons. High-potential community innovators are blocked by the technical barrier of entry.

## Solution Overview
ImpactFlow acts as an **Autonomous Technical Consultant**. It doesn't just listen; it critiques, analyzes, metrics, and blueprints. It uses Gemini to turn a sentence into a 5-phase engineering roadmap.

## Key Features
- **Categorical Impact Analysis:** Extracts domain insights using zero-shot classification.
- **Dynamic Architecture Rendering:** Uses Mermaid.js to "see" the implementation.
- **Educational Tooltips:** Simplifies senior-level technical jargon into "Student Tips."
- **Leaderboard of Good:** Gamifies community problem identification.

## Technologies Used
- **Google Gemini 2.0 Flash:** The engine.
- **Typescript/Node:** The structure.
- **Mermaid.js:** The visual logic.

## Target Users
- **Middle/High School Students:** Specifically Girls in STEM competing in social good hackathons.
- **Hackathon Organizers:** To help students reach the "Prototype" stage faster.

[impactflow.zip](https://github.com/user-attachments/files/27763791/impactflow.zip)
[ImpactFlow_Presentation_Report.pptx](https://github.com/user-attachments/files/27763790/ImpactFlow_Presentation_Report.pptx)
